src/MongoDB.EntityFrameworkCore/Storage/MongoDatabaseWrapper.cs

```diff
@@ -238,57 +223,67 @@ private static MongoUpdate ConvertModifiedEntryToMongoUpdate(string collectionNa
         var entityType = entry.EntityType;
         var entitySerializer = (IBsonDocumentSerializer)EntitySerializer.Create(entityType);
 
-        var idProperty = entityType.GetIdProperty() ?? throw new InvalidOperationException($"Type {entityType.ClrType.Name} has no Id property.");
-        var idValue = entry.GetCurrentValue(idProperty);
-        var idSerializationInfo = GetPropertySerializationInfo(entitySerializer, idProperty);
-        var idSerializer = idSerializationInfo.Serializer;
-        var serializedIdValue = SerializeValue(idSerializer, idValue);
-        var filterDefinition = Builders<BsonDocument>.Filter.Eq("_id", serializedIdValue);
-
         var fieldValues = new BsonDocument();
         foreach (var property in entityType.GetProperties())
         {
             if (entry.IsModified(property))
             {
                 var propertyValue = entry.GetCurrentValue(property);
                 var propertySerializationInfo = GetPropertySerializationInfo(entitySerializer, property);
-                var elementName = propertySerializationInfo.ElementName;
                 var propertySerializer = propertySerializationInfo.Serializer;
                 var serializedPropertyValue = SerializeValue(propertySerializer, propertyValue);
+
+                var elementName = propertySerializationInfo.ElementName;
                 fieldValues[elementName] = serializedPropertyValue;
             }
         }
+
         var updateDocument = new BsonDocument("$set", fieldValues);
         var updateDefinition = new BsonDocumentUpdateDefinition<BsonDocument>(updateDocument);
 
-        var model = new UpdateOneModel<BsonDocument>(filterDefinition, updateDefinition);
+        var idFilter = CreateIdFilter(entitySerializer, entry);
+        var model = new UpdateOneModel<BsonDocument>(idFilter, updateDefinition);
         return new MongoUpdate(collectionName, model);
+    }
 
-        static BsonSerializationInfo GetPropertySerializationInfo(IBsonDocumentSerializer entitySerializer, IReadOnlyProperty property)
-        {
-            if (entitySerializer.TryGetMemberSerializationInfo(property.Name, out var serializationInfo))
-            {
-                return serializationInfo;
-            }
+    private static FilterDefinition<BsonDocument> CreateIdFilter(IBsonDocumentSerializer serializer, IUpdateEntry entry)
+    {
+        var idProperty = entry.EntityType.GetIdProperty() ??
```

Comment:
Should `GetIdProperty` throw an exception if it can't find the id property?

That way callers don't have to check.

I think if we call `GetIdProperty` it's because we need to have it. I can't think of a case where we would want to get `null` back here.